packages/nx/src/command-line/exec/exec.ts

```diff
@@ -117,26 +153,40 @@ function ensureNxTarget(project: ProjectGraphProjectNode, targetName: string) {
   }
 }
 
-function getProject(projectGraph: ProjectGraph) {
-  const projectName = calculateDefaultProjectName(
+function getProjects(
+  projectGraph: ProjectGraph,
+  nxArgs: NxArgs
+): ProjectGraphProjectNode[] {
+  let selectedProjects = {};
+
+  const defaultProjectName = calculateDefaultProjectName(
     process.cwd(),
     workspaceRoot,
     readProjectsConfigurationFromProjectGraph(projectGraph),
     readNxJson()
   );
+  if (defaultProjectName && projectGraph.nodes[defaultProjectName]) {
+    selectedProjects[defaultProjectName] =
+      projectGraph.nodes[defaultProjectName];
+  }
 
-  if (!projectName) {
-    output.error({
-      title: 'Unable to determine project name for `nx exec`',
-      bodyLines: [
-        "`nx exec` should be ran from within an Nx project's root directory.",
-        'Does this package.json belong to an Nx project?',
-      ],
-    });
-    process.exit(1);
+  // get projects matched
+  if (nxArgs.projects?.length) {
+    for (const projectName of nxArgs.projects) {
+      selectedProjects[projectName] = projectGraph.nodes[projectName];
+    }
+  } else {
+    // if no project specified, return all projects
+    for (const projectName in projectGraph.nodes) {
+      selectedProjects[projectName] = projectGraph.nodes[projectName];
+    }
```

Comment:
Wouldn't this mean that if you have `nx exec -- tsc` in a package.json's build script, it would try to run tsc on every project by default? We don't want to change existing behavior of the command, just add on to it right?

Reply:
i added an e2e test for this. i think when running `npm run` or `nx run`, it would set the env NX_TASK_TARGET_PROJECT, which will exec the command after -- directly.

Reply:
That's true for `nx run`, `npm run` doesn't set any kind of NX_... variables until nx is running though

Reply:
ok. i see. i change the logic whereas if process.env.npm_lifecycle_event is defined, just use the default project.

Reply:
Looks good 🎉